go/enclave/rollupmanager/rollupmanager.go

```diff
@@ -0,0 +1,286 @@
+package rollupmanager
+
+// TODO: once the cross chain messages based bridge is implemented remove this completely.
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+	"sort"
+
+	gethlog "github.com/ethereum/go-ethereum/log"
+
+	"github.com/obscuronet/go-obscuro/contracts/generated/MessageBus"
+	"github.com/obscuronet/go-obscuro/go/common"
+	"github.com/obscuronet/go-obscuro/go/common/errutil"
+	"github.com/obscuronet/go-obscuro/go/common/log"
+
+	"github.com/obscuronet/go-obscuro/go/ethadapter"
+
+	crypto2 "github.com/obscuronet/go-obscuro/go/enclave/crypto"
+	"github.com/obscuronet/go-obscuro/go/enclave/l2chain"
+
+	gethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/obscuronet/go-obscuro/go/enclave/core"
+	"github.com/obscuronet/go-obscuro/go/enclave/db"
+	"github.com/obscuronet/go-obscuro/go/ethadapter/mgmtcontractlib"
+)
+
+// RollupManager encapsulates the logic of decoding rollup transactions submitted to the L1 and resolving them
+// to rollups that the enclave can process.
+type RollupManager struct {
+	MgmtContractLib mgmtcontractlib.MgmtContractLib
+
+	TransactionBlobCrypto crypto2.TransactionBlobCrypto
+
+	ObscuroChainID  int64
+	EthereumChainID int64
+
+	logger gethlog.Logger
+
+	l2chain *l2chain.ObscuroChain
+	storage db.Storage
+}
+
+func New(
+	mgmtContractLib mgmtcontractlib.MgmtContractLib,
+	transactionBlobCrypto crypto2.TransactionBlobCrypto,
+	obscuroChainID int64,
+	ethereumChainID int64,
+	storage db.Storage,
+	l2chain *l2chain.ObscuroChain,
+	logger gethlog.Logger,
+) *RollupManager {
+	return &RollupManager{
+		MgmtContractLib:       mgmtContractLib,
+		TransactionBlobCrypto: transactionBlobCrypto,
+		ObscuroChainID:        obscuroChainID,
+		EthereumChainID:       ethereumChainID,
+		logger:                logger,
+		l2chain:               l2chain,
+		storage:               storage,
+	}
+}
+
+func (re *RollupManager) fetchLatestRolluп() (*core.Rollup, error) {
+	b, err := re.storage.FetchHeadBlock()
+	if err != nil {
+		return nil, err
+	}
+	return re.getLatestRollupBeforeBlock(b)
+}
+
+func NextRollup(rollup *core.Rollup, batches []*core.Batch) *core.Rollup {
+	headBatch := batches[len(batches)-1]
+
+	rh := headBatch.Header.ToRollupHeader()
+
+	if rollup != nil {
+		rh.ParentHash = rollup.Header.Hash()
+	} else { // genesis
+		rh.ParentHash = gethcommon.Hash{}
+	}
+
+	rh.CrossChainMessages = make([]MessageBus.StructsCrossChainMessage, 0)
```

Comment:
The slice might not be nil since im still using the batch header promotion which might contain messages. But it is the last batch header so its messages must come last, rather than first and then again at the back, thus the new array.